upf/io/pddl_reader.py

```diff
@@ -14,24 +14,321 @@
+        self._em = self._env.expression_manager
+        self._tm = self._env.type_manager
+        self._operators = {
+            'and' : self._em.And,
+            'or' : self._em.Or,
+            'not' : self._em.Not,
+            'imply' : self._em.Implies,
+            '>=' : self._em.GE,
+            '<=' : self._em.LE,
+            '>' : self._em.GT,
+            '<' : self._em.LT,
+            '=' : self._em.Equals,
+            '+' : self._em.Plus,
+            '-' : self._em.Minus,
+            '/' : self._em.Div,
+            '*' : self._em.Times
+        }
+
+    def _parse_exp(self, problem, act, var, exp):
+        if isinstance(exp, ParseResults) and len(exp) == 0:
+            return self._em.TRUE()
+        elif isinstance(exp, ParseResults) and exp[0] == '-' and len(exp) == 2:
+            return Times(-1, self._parse_exp(problem, act, var, exp[1]))
+        elif isinstance(exp, ParseResults) and exp[0] in self._operators:
+            op = self._operators[exp[0]]
+            exp = exp[1:]
+            return op(*[self._parse_exp(problem, act, var, e) for e in exp])
+        elif isinstance(exp, ParseResults) and exp[0] in ['exists', 'forall']:
+            op = self._em.Exists if exp[0] == 'exists' else self._em.Forall
+            vars_string = ' '.join(exp[1])
+            vars_res = parameters.parseString(vars_string)
+            vars = {}
+            for g in vars_res['params']:
+                t = self._tm.UserType(g[1] if len(g) > 1 else 'object')
+                for o in g[0]:
+                    vars[o] = upf.model.Variable(o, t)
+            return op(self._parse_exp(problem, act, vars, exp[2]), *vars.values())
+        elif isinstance(exp, ParseResults) and problem.has_fluent(exp[0]):
+            f = problem.fluent(exp[0])
+            args = [self._parse_exp(problem, act, var, e) for e in exp[1:]]
+            return self._em.FluentExp(f, tuple(args))
+        elif isinstance(exp, ParseResults) and len(exp) == 1:
+            return self._parse_exp(problem, act, var, exp[0])
+        elif isinstance(exp, str) and exp[0] == '?' and exp[1:] in var:
+            return self._em.VariableExp(var[exp[1:]])
+        elif isinstance(exp, str) and exp[0] == '?':
+            return self._em.ParameterExp(act.parameter(exp[1:]))
+        elif isinstance(exp, str) and problem.has_fluent(exp):
+            return self._em.FluentExp(problem.fluent(exp))
+        elif isinstance(exp, str) and problem.has_object(exp):
+            return self._em.ObjectExp(problem.object(exp))
+        elif isinstance(exp, str):
+            return self._em.Real(Fraction(exp))
+        else:
+            raise upf.exceptions.UPFProblemDefinitionError(f'Not able to handle: {exp}')
+
+    def _add_effect(self, problem, act, exp, cond=True, timing=None):
+        op = exp[0]
+        if op == 'and':
+            exp = exp[1:]
+            for e in exp:
+                self._add_effect(problem, act, e, cond, timing)
```

Comment:
It should be very easy to remove recursion here: it suffices to implement a super-basic conjunctive partitioning like: https://github.com/pysmt/pysmt/blob/master/pysmt/rewritings.py#L856